### ISSUE TITLE
networkd - architecture matches on invalid values

### ISSUE THREAD

**Issue by XeS0r** [systemd/systemd#3236]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

229+ / Upstream

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### In case of bug report: Expected behaviour you didn't see

Files with invalid architecture values should be skipped/ignored.
### In case of bug report: Unexpected behaviour you saw

Using invalid values for architecture leads to a match instead of a mismatch. I've only tested netdev files but since this is used across different config files it should affect all which are using architecture in their match section.
### In case of bug report: Steps to reproduce the problem

Create NetDev file with content:

```
[Match]
Architecture=**INVALID**

[NetDev]
Description=NetDev configuration for testing purpose
Name=dummyinf0
Kind=dummy
```

As far as I could track it down it seems that `architecture_from_string`(condition.c line 186) is returning `_ARCHITECTURE_INVALID = -1` ([architecture.h line 62](https://github.com/systemd/systemd/blob/master/src/basic/architecture.h#L62)). Which then will be returned by `condition_test_architecture`([condition.c line 188](https://github.com/systemd/systemd/blob/master/src/shared/condition.c#L188)) to indicate an error. But later this return value is interpreted as true in `net_match_config`([network-internal.c line 113](https://github.com/systemd/systemd/blob/master/src/libsystemd-network/network-internal.c#L113)).

`if (match_arch && !condition_test(match_arch))`

potential solution
`if (match_arch && condition_test(match_arch) <= 0)`

But this won't hint the user that an invalid value has been supplied in his configuration.

This issue is not only related to architecture. It could happen to others to if a negative value is returned by the condition function.

e.g.: If `detect_virtualization()` call in `condition_test_virtualization`([condition.c line 141](https://github.com/systemd/systemd/blob/master/src/shared/condition.c#L141)) returns a negative value.

**Comment by evverx** [comment]
> potential solution
> if (match_arch && condition_test(match_arch) <= 0)
> 
> But this won't hint the user that an invalid value has been supplied in his configuration.

Indeed.
Maybe, we need something like this

``` c
                r = condition_test(c);
                if (r < 0)
                        log_unit_warning(u,
                                         "Couldn't determine result for %s=%s%s%s, assuming failed: %m",
                                         to_string(c->type),
                                         c->trigger ? "|" : "",
                                         c->negate ? "!" : "",
                                         c->parameter);
                else
                        log_unit_debug(u,
                                       "%s=%s%s%s %s.",
                                       to_string(c->type),
                                       c->trigger ? "|" : "",
                                       c->negate ? "!" : "",
                                       c->parameter,
                                       condition_result_to_string(c->result));

```

https://github.com/systemd/systemd/blob/master/src/core/unit.c#L1323

**Comment by poettering** [comment]
Fix waiting in #3633.
